### PR TITLE
[SQL][SPARK-6742]: Don't push down predicates which reference partition column(s)

### PR DESCRIPTION
**Pull request by saucam**
cc @liancheng

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 1]
```diff
@@ -214,6 +214,12 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
       case logical.InsertIntoTable(table: ParquetRelation, partition, child, overwrite) =>
         InsertIntoParquetTable(table, planLater(child), overwrite) :: Nil
       case PhysicalOperation(projectList, filters: Seq[Expression], relation: ParquetRelation) =>
+        val partitionColNames = relation.partitioningAttributes.map(_.name).toSet
+        val filtersToPush = filters
+          .filter { pred =>
+            val referencedColNames = pred.references.map(_.name).toSet
+            referencedColNames.intersect(partitionColNames).isEmpty
+          }
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Nit, we'd prefer the following style:

```
val filtersToPush = filters.filter { pred =>
  ...
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 2]
```diff
@@ -223,8 +229,12 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
               // can result in "A OR C" being pushed down. Here we are conservative in the sense
               // that even if "A" was pushed and we check for "A AND B" we still want to keep
               // "A AND B" in the higher-level filter, not just "B".
-              predicates.map(p => p -> ParquetFilters.createFilter(p)).collect {
+              predicates.map(p => p -> ParquetFilters.createFilter(p)).collect {
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Remove the trailing space please.

**Comment by liancheng** [comment]
Good catch! #5210 fixed this for the new Parquet data source but forgot the old Parquet code path. Would you please to add a test case for this? Thanks!

**Comment by saucam** [comment]
Added test case. please test

**Comment by marmbrus** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetFilterSuite.scala [hunk 3]
```diff
@@ -22,7 +22,8 @@ import parquet.filter2.predicate.Operators._
 import parquet.filter2.predicate.{FilterPredicate, Operators}
 
 import org.apache.spark.sql.catalyst.dsl.expressions._
-import org.apache.spark.sql.catalyst.expressions.{Attribute, Cast, Literal, Predicate, Row}
+import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference, Cast,
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetFilterSuite.scala]: [Don't wrap import statements](https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-LineLength).  After 6+ imports consider using `_`.

**Comment by marmbrus** [comment]
Thanks!  Merged to master.

**Pull request closed without merging**